streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java

```diff
@@ -422,440 +439,268 @@ public void shouldNotThrowStreamsExceptionOnSubsequentCallIfASendFailsWithContin
+        final StreamsException thrown = assertThrows(
+            StreamsException.class,
+            () -> new RecordCollectorImpl(
                 logContext,
-                streamsMetrics,
-                null,
-                id -> new MockProducer<byte[], byte[]>(cluster, true, new DefaultPartitioner(), byteArraySerializer, byteArraySerializer) {
-                    @Override
-                    public void initTransactions() {
-                        throw exception;
-                    }
-                }
+                taskId,
+                mockConsumer,
+                new StreamsProducer(mockProducer, taskId),
+                productionExceptionHandler,
+                "appId",
+                true,
+                streamsMetrics
             )
         );
-        assertEquals(exception, thrown.getCause());
+
+        assertEquals(mockProducer.initTransactionException, thrown.getCause());
+        assertThat(thrown.getMessage(), equalTo("Timeout exception caught when initializing transactions for task 0_0. \nThe broker is either slow or in bad state (like not having enough replicas) in responding to the request, or the connection to broker was interrupted sending the request or receiving the response. \n Consider overwriting `max.block.ms` to a larger value to avoid timeout errors"));
     }
 
     @Test
     public void shouldThrowStreamsExceptionOnEOSInitializeError() {
-        final KafkaException exception = new KafkaException("KABOOM!");
-        final Properties props = StreamsTestUtils.getStreamsConfig("test");
-        props.setProperty(StreamsConfig.PROCESSING_GUARANTEE_CONFIG, StreamsConfig.EXACTLY_ONCE);
+        // use `mockProducer` instead of `eosMockProducer` to avoid double Tx-Init
+        mockProducer.initTransactionException = new KafkaException("KABOOM!");
 
-        final StreamsException thrown = assertThrows(StreamsException.class, () ->
-            new RecordCollectorImpl(
-                taskId,
-                new StreamsConfig(props),
+        final StreamsException thrown = assertThrows(
+            StreamsException.class,
+            () -> new RecordCollectorImpl(
                 logContext,
-                streamsMetrics,
-                null,
-                id -> new MockProducer<byte[], byte[]>(cluster, true, new DefaultPartitioner(), byteArraySerializer, byteArraySerializer) {
-                    @Override
-                    public void initTransactions() {
-                        throw exception;
-                    }
-                }
+                taskId,
+                mockConsumer,
+                new StreamsProducer(mockProducer, taskId),
+                productionExceptionHandler,
+                "appId",
+                true,
+                streamsMetrics
             )
         );
-        assertEquals(exception, thrown.getCause());
-    }
 
-    @Test
-    public void shouldThrowMigrateExceptionOnEOSFirstSendProducerFenced() {
```

Comment:
The name and the test does not really match as it throws in `beginTransaction`. In any case, we have test for both cases already and thus I removed this duplicate test (please verify that I did not reduce test coverage)

Reply:
My argument was that the old test was named "send" but did throw in "beginTx" (and there was already a corresponding "beginTx" test in the old code -> hence "redundant" / or at least need a fix). However, the new `StreamsProducerTest` should cover all cases for "send" as well as "beginTx":

BeginTx:
- `shouldThrowTaskMigrateExceptionOnEosBeginTxnFenced`
- `shouldThrowTaskMigrateExceptionOnEosBeginTxnError`
- `shouldFailOnEosBeginTxnFatal`

send:
 - `shouldThrowTaskMigratedExceptionOnEosSendFenced`
 - `shouldThrowTaskMigratedExceptionOnEosSendUnknownPid`
 - `shouldThrowTaskMigrateExceptionOnEosSendOffsetFenced`
 - `shouldThrowStreamsExceptionOnEosSendOffsetError`
 - `shouldFailOnEosSendOffsetFatal`